src/moonlink/src/storage/wal.rs

```diff
@@ -1 +1,331 @@
+            TableEvent::StreamFlush { xact_id } => WalEventEnum::StreamFlush { xact_id: *xact_id },
+            _ => {
+                unimplemented!("Invalid table event for WAL: {:?}", table_event)
+            }
+        };
+
+        Self { lsn, event }
+    }
+
+    pub fn into_table_event(self) -> TableEvent {
+        match self.event {
+            WalEventEnum::Append {
+                row,
+                xact_id,
+                is_copied,
+            } => TableEvent::Append {
+                row,
+                xact_id,
+                is_copied,
+                lsn: self.lsn,
+            },
+            WalEventEnum::Delete { row, xact_id } => TableEvent::Delete {
+                row,
+                xact_id,
+                lsn: self.lsn,
+            },
+            WalEventEnum::Commit { xact_id } => TableEvent::Commit {
+                xact_id,
+                lsn: self.lsn,
+            },
+            WalEventEnum::StreamAbort { xact_id } => TableEvent::StreamAbort { xact_id },
+            WalEventEnum::StreamFlush { xact_id } => TableEvent::StreamFlush { xact_id },
+        }
+    }
+}
+struct InMemWal {
+    /// The in_mem_wal could have insertions done by incoming CDC events
+    pub buf: Vec<WalEvent>,
+    /// Tracks an LSN in case of a stream flush (or similar events) which has no accompanying LSN.
+    /// A new instance will take the highest_lsn from the previous instance before it is populated.
+    pub highest_lsn: u64,
+}
+
+impl InMemWal {
+    pub fn new(highest_lsn: u64) -> Self {
+        Self {
+            buf: Vec::new(),
+            highest_lsn,
+        }
+    }
+}
+
+/// Wal tracks both the in-memory WAL and the flushed WALs.
+/// Note that wal manager is meant to be used in a single thread. While
+/// persist and delete_files can be called concurrently, their results
+/// have to be handled serially.
+pub(crate) struct WalManager {
+    in_mem_wal: InMemWal,
+    /// The wal file numbers that are still live.
+    live_wal_file_tracker: Vec<WalFileInfo>,
+    /// Tracks the file number to be assigned to the next persisted wal file
+    curr_file_number: u64,
+
+    file_system_accessor: Arc<dyn BaseFileSystemAccess>,
+}
+
+impl WalManager {
+    pub fn new(config: FileSystemConfig) -> Self {
+        // TODO(Paul): Add a more robust constructor when implementing recovery
+        Self {
+            in_mem_wal: InMemWal::new(0),
+            live_wal_file_tracker: Vec::new(),
+            curr_file_number: 0,
+            file_system_accessor: Arc::new(FileSystemAccessor::new(config)),
+        }
+    }
+
+    pub fn insert(&mut self, table_event: &TableEvent) {
+        let wal_event = WalEvent::new(table_event, self.in_mem_wal.highest_lsn);
+        self.in_mem_wal.buf.push(wal_event);
+
+        // Update highest_lsn if this event has a higher LSN
+        if let Some(lsn) = table_event.get_lsn_for_ingest_event() {
+            assert!(lsn >= self.in_mem_wal.highest_lsn);
```

Comment:
```suggestion
            ma::assert_ge!(lsn, self.in_mem_wal.highest_lsn);
```
which prints out value if assertion fails